contracts/src/token/erc721/extensions/consecutive.rs

```diff
@@ -0,0 +1,1079 @@
+    event ConsecutiveTransfer(
+        uint256 indexed fromTokenId,
+        uint256 toTokenId,
+        address indexed fromAddress,
+        address indexed toAddress
+    );
+}
+
+sol! {
+    /// Batch mint is restricted to the constructor.
+    /// Any batch mint not emitting the [`IERC721::Transfer`] event outside of the constructor
+    /// is non ERC-721 compliant.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC721ForbiddenBatchMint();
+
+    /// Exceeds the max number of mints per batch.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC721ExceededMaxBatchMint(uint256 batchSize, uint256 maxBatch);
+
+    /// Individual minting is not allowed.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC721ForbiddenMint();
+
+    /// Batch burn is not supported.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC721ForbiddenBatchBurn();
+}
+
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// Error type from erc721 contract [`erc721::Error`]
+    Erc721(erc721::Error),
+    /// Error type from checkpoint contract [`checkpoints::Error`]
+    Checkpoints(checkpoints::Error),
+    /// Batch mint is restricted to the constructor.
+    /// Any batch mint not emitting the [`IERC721::Transfer`] event outside of
+    /// the constructor is non ERC-721 compliant.
+    ForbiddenBatchMint(ERC721ForbiddenBatchMint),
+    /// Exceeds the max amount of mints per batch.
+    ExceededMaxBatchMint(ERC721ExceededMaxBatchMint),
+    /// Individual minting is not allowed.
+    ForbiddenMint(ERC721ForbiddenMint),
+    /// Batch burn is not supported.
+    ForbiddenBatchBurn(ERC721ForbiddenBatchBurn),
+}
+
+impl MethodError for erc721::Error {
+    fn encode(self) -> alloc::vec::Vec<u8> {
+        self.into()
+    }
+}
+
+impl MethodError for checkpoints::Error {
+    fn encode(self) -> alloc::vec::Vec<u8> {
+        self.into()
+    }
+}
+
+// TODO: add an option to override these constants
+
+/// Maximum size of a batch of consecutive tokens. This is designed to limit
+/// stress on off-chain indexing services that have to record one entry per
+/// token, and have protections against "unreasonably large" batches of tokens.
+pub const MAX_BATCH_SIZE: U96 = uint!(5000_U96);
+
+/// Used to offset the first token id in
+/// [`Erc721Consecutive::_next_consecutive_id`]
+pub const FIRST_CONSECUTIVE_ID: U96 = uint!(0_U96);
+
+/// Consecutive extension related implementation:
+impl Erc721Consecutive {
+    /// Override of [`Erc721::_owner_of_inner`] that checks the sequential
+    /// ownership structure for tokens that have been minted as part of a
+    /// batch, and not yet transferred.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `token_id` - Token id as a number.
+    pub fn _owner_of_inner(&self, token_id: U256) -> Address {
+        let owner = self.__owner_of_inner(token_id);
```

Comment:
Why not inline `__owner_of_inner` to `self.erc721._owner_of_inner`? 

Also, I just realized that we don't need to suffix `_inner` in the Erc721 implementation.